File: clients/php/Model/VehicleType.php
<?php

namespace Model {
    require_once 'Stream.php';

    /**
     * Vehicle type options
     */
    class VehicleType
    {
        /**
         * Name
         */
        public string $name;
        /**
         * Radius
         */
        public float $radius;
        /**
         * Weight
         */
        public float $weight;
        /**
         * Maximal backwads movement speed
         */
        public float $maxBackwardsSpeed;
        /**
         * Maximal forward movement speed
         */
        public float $maxSpeed;
        /**
         * Acceleration
         */
        public float $acceleration;
        /**
         * Friction coefficient
         */
        public float $friction;
        /**
         * Maximal rotation speed
         */
        public float $maxRotateSpeed;
        /**
         * Rotational acceleration
         */
        public float $rotateAcceleration;
        /**
         * Maximal amount of fuel
         */
        public float $maxFuel;
        /**
         * Fuel usage speed
         */
        public float $fuelUseSpeed;
    
        function __construct(string $name, float $radius, float $weight, float $maxBackwardsSpeed, float $maxSpeed, float $acceleration, float $friction, float $maxRotateSpeed, float $rotateAcceleration, float $maxFuel, float $fuelUseSpeed)
        {
            $this->name = $name;
            $this->radius = $radius;
            $this->weight = $weight;
            $this->maxBackwardsSpeed = $maxBackwardsSpeed;
            $this->maxSpeed = $maxSpeed;
            $this->acceleration = $acceleration;
            $this->friction = $friction;
            $this->maxRotateSpeed = $maxRotateSpeed;
            $this->rotateAcceleration = $rotateAcceleration;
            $this->maxFuel = $maxFuel;
            $this->fuelUseSpeed = $fuelUseSpeed;
        }
    
        /**
         * Read VehicleType from input stream
         */
        public static function readFrom(\InputStream $stream): VehicleType
        {
            $name = $stream->readString();
            $radius = $stream->readDouble();
            $weight = $stream->readDouble();
            $maxBackwardsSpeed = $stream->readDouble();
            $maxSpeed = $stream->readDouble();
            $acceleration = $stream->readDouble();
            $friction = $stream->readDouble();
            $maxRotateSpeed = $stream->readDouble();
            $rotateAcceleration = $stream->readDouble();
            $maxFuel = $stream->readDouble();
            $fuelUseSpeed = $stream->readDouble();
            return new VehicleType($name, $radius, $weight, $maxBackwardsSpeed, $maxSpeed, $acceleration, $friction, $maxRotateSpeed, $rotateAcceleration, $maxFuel, $fuelUseSpeed);
        }
        
        /**
         * Write VehicleType to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeString($this->name);
            $stream->writeDouble($this->radius);
            $stream->writeDouble($this->weight);
            $stream->writeDouble($this->maxBackwardsSpeed);
            $stream->writeDouble($this->maxSpeed);
            $stream->writeDouble($this->acceleration);
            $stream->writeDouble($this->friction);
            $stream->writeDouble($this->maxRotateSpeed);
            $stream->writeDouble($this->rotateAcceleration);
            $stream->writeDouble($this->maxFuel);
            $stream->writeDouble($this->fuelUseSpeed);
        }
    }
}

File: clients/cpp/model/VehicleType.cpp
#include "VehicleType.hpp"

namespace model {

VehicleType::VehicleType(std::string name, double radius, double weight, double maxBackwardsSpeed, double maxSpeed, double acceleration, double friction, double maxRotateSpeed, double rotateAcceleration, double maxFuel, double fuelUseSpeed) : name(name), radius(radius), weight(weight), maxBackwardsSpeed(maxBackwardsSpeed), maxSpeed(maxSpeed), acceleration(acceleration), friction(friction), maxRotateSpeed(maxRotateSpeed), rotateAcceleration(rotateAcceleration), maxFuel(maxFuel), fuelUseSpeed(fuelUseSpeed) { }

// Read VehicleType from input stream
VehicleType VehicleType::readFrom(InputStream& stream) {
    std::string name = stream.readString();
    double radius = stream.readDouble();
    double weight = stream.readDouble();
    double maxBackwardsSpeed = stream.readDouble();
    double maxSpeed = stream.readDouble();
    double acceleration = stream.readDouble();
    double friction = stream.readDouble();
    double maxRotateSpeed = stream.readDouble();
    double rotateAcceleration = stream.readDouble();
    double maxFuel = stream.readDouble();
    double fuelUseSpeed = stream.readDouble();
    return VehicleType(name, radius, weight, maxBackwardsSpeed, maxSpeed, acceleration, friction, maxRotateSpeed, rotateAcceleration, maxFuel, fuelUseSpeed);
}

// Write VehicleType to output stream
void VehicleType::writeTo(OutputStream& stream) const {
    stream.write(name);
    stream.write(radius);
    stream.write(weight);
    stream.write(maxBackwardsSpeed);
    stream.write(maxSpeed);
    stream.write(acceleration);
    stream.write(friction);
    stream.write(maxRotateSpeed);
    stream.write(rotateAcceleration);
    stream.write(maxFuel);
    stream.write(fuelUseSpeed);
}

// Get string representation of VehicleType
std::string VehicleType::toString() const {
    std::stringstream ss;
    ss << "VehicleType { ";
    ss << "name: ";
    ss << '"' << name << '"';
    ss << ", ";
    ss << "radius: ";
    ss << radius;
    ss << ", ";
    ss << "weight: ";
    ss << weight;
    ss << ", ";
    ss << "maxBackwardsSpeed: ";
    ss << maxBackwardsSpeed;
    ss << ", ";
    ss << "maxSpeed: ";
    ss << maxSpeed;
    ss << ", ";
    ss << "acceleration: ";
    ss << acceleration;
    ss << ", ";
    ss << "friction: ";
    ss << friction;
    ss << ", ";
    ss << "maxRotateSpeed: ";
    ss << maxRotateSpeed;
    ss << ", ";
    ss << "rotateAcceleration: ";
    ss << rotateAcceleration;
    ss << ", ";
    ss << "maxFuel: ";
    ss << maxFuel;
    ss << ", ";
    ss << "fuelUseSpeed: ";
    ss << fuelUseSpeed;
    ss << " }";
    return ss.str();
}

}